velox/functions/prestosql/window/Ntile.cpp

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/common/base/Exceptions.h"
+#include "velox/exec/WindowFunction.h"
+#include "velox/expression/FunctionSignature.h"
+#include "velox/vector/FlatVector.h"
+
+namespace facebook::velox::window {
+
+namespace {
+
+class NtileFunction : public exec::WindowFunction {
+ public:
+  explicit NtileFunction(const std::vector<exec::WindowFunctionArg>& args)
+      : WindowFunction(BIGINT(), nullptr) {
+    auto argBuckets = args[0].constantValue;
+    VELOX_CHECK_NOT_NULL(argBuckets, kBucketErrorString);
+    VELOX_USER_CHECK(!argBuckets->isNullAt(0), kBucketErrorString);
+    numBuckets_ = argBuckets->as<ConstantVector<int64_t>>()->valueAt(0);
+    VELOX_USER_CHECK(numBuckets_ >= 1, kBucketErrorString);
+  }
+
+  void resetPartition(const exec::WindowPartition* partition) {
+    partitionOffset_ = 0;
+    numPartitionRows_ = partition->numRows();
+
+    hasRemainderRows_ = numPartitionRows_ > numBuckets_;
```

Comment:
The names of the variables are confusing to me.

hasRemainderRows_ - I'd expect this to be true if remainderRows_ > 0 and false if remainderRows_ == 0.

Also, rowsPerBucket_ is strange. When we have 6 rows and 4 buckets, rowsPerBucket_ is 1 even though the result has 2 rows per bucket: 1, 1, 2, 2, 3, 4.

Also, I'm seeing that number of rows in a bucket is not fixed. For example, with 20 rows and 9 buckets we get 2 buckets of 3 rows and 7 buckets of 2 rows.

```
presto:di> with t as (select * from unnest(sequence(0, 19)) as _t(x))
        -> SELECT x, ntile(9) over (partition by 'fixed')
        -> FROM t
        -> ;
 x  | _col1
----+-------
  0 |     1
  1 |     1
  2 |     1
  3 |     2
  4 |     2
  5 |     2
  6 |     3
  7 |     3
  8 |     4
  9 |     4
 10 |     5
 11 |     5
 12 |     6
 13 |     6
 14 |     7
 15 |     7
 16 |     8
 17 |     8
 18 |     9
 19 |     9
(20 rows)
```

Also, I'm seeing Presto allowing non-constant inputs to ntile:

```
presto:di> with t as (select * from unnest(sequence(0, 4)) as _t(x))
        -> SELECT x, ntile(x + 1) over (partition by 'fixed')
        -> FROM t
        -> ;
 x | _col1
---+-------
 0 |     1
 1 |     1
 2 |     2
 3 |     3
 4 |     5
(5 rows)
```

Reply:
Presto does indeed allow non-constant inputs. I had written the code this way only to demonstrate the use of the constant parameter. But intended to enhance it in the future. 

Have now modified the code to handle column parameters and have also changed variable naming, more comments etc to explain the behavior you have seen above.

Would be great to get another round of review on the new code.